drivers/i2c/i2c_smartbond.c

```diff
@@ -31,8 +33,41 @@ struct i2c_smartbond_data {
 	uint32_t transmit_cnt, receive_cnt;
 	i2c_callback_t cb;
 	void *userdata;
+#if defined(CONFIG_PM_DEVICE) || defined(CONFIG_PM_DEVICE_RUNTIME)
+	/* Flag to designate whether or not a transaction is in progress */
+	volatile bool in_progress;
```

Comment:
Couldn't this be done by informing pm the device is busy instead? https://docs.zephyrproject.org/latest/services/pm/api/index.html#c.pm_device_busy_set

Maybe broader, why is a custom get/put pair even needed here

Reply:
Perhaps i2c_pm_get/put should be renamed to avoid confusion with runtime pm put/get.

Reply:
Functions have been renamed as suggested

Reply:
> A. This scheme should not work in our case as when the SoC enters the sleep state, it's completely turned off. The scheme we have chosen is that, as long as there is some kind of peripheral activity a negative error code is returned by the registered PM callback function. This should result in aborting the sleep state and entering the WFI state, instead.

this is exactly what get and put are there for, they are reference counts. When the ref count is > 0 the pm callback shouldn't even be called to put the device to sleep.

https://docs.zephyrproject.org/latest/services/pm/device_runtime.html#implementation-guidelines

@ceolin please take a look at correct me if I'm wrong here, but this looks odd

Reply:
@teburd actually when device runtime is enabled for a device the system power management won't suspend/resume the device, it does not matter its reference count. The problem (if I understood correctly) is that when the SoC is suspended (probably depending which power state is used) the device lose power. It is much likely an internal device and the power domain where it is is gated when the SoC sleeps.

That is said, return an ERROR to avoid the SoC to be suspended is not the best approach, it wastes a  lot of resources. The proper solution would be to set a power constraints in the power state that you don't want the system to use. See https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/pm/policy.h#L114

I believe that may be a good use case for this pr as well https://github.com/zephyrproject-rtos/zephyr/pull/70111
It introduces a way to declare in DT which power states would cause power loss for a device and adds an api to automatically set and release constraints in them. https://github.com/zephyrproject-rtos/zephyr/pull/70111/commits/cb2e759a965139934c2fb9e9fd732382393ecfb6#diff-805a1b927cb7a1366937061819faed4046a1ea3472476fc342e637dee847dc47

You would call  `pm_policy_device_power_lock_get()` when you are busy and ` pm_policy_device_power_lock_put` when finished the work. This avoids the whole work that happens when you return an error there.

That obviously does not solve the problem of someone calling `pm_device_runtime_put` while you are in the middle of a transaction. For that problem the device should be calling `pm_device_runtime_get(self)`  and `pm_device_runtime_put(self)` to not let the reference count gets 0.

@ioannis-karachalios let me know if that makes sense.

Reply:
@ceolin thanks for the thorough analysis provided above. Indeed, returning an error code should not be optimized and the best approach is the usage of power constraints when the device should not enter a sleep state (e.g. when a transaction is ongoing). 
As far as https://github.com/zephyrproject-rtos/zephyr/pull/70111 is concerned, this is quite handful especially if multiple power states should be constrained or node instances should not behave at exactly the same way. In our case we only support a single power state. 
With regards to your last statement, I think this could be omitted if there is provision for the device driver to wait till an ongoing transaction is done and/or the refutation of power constrains is done at the right place.

Reply:
Thanks for the informative review @ceolin